src/Microsoft.Extensions.Primitives/StringValues.cs

```diff
@@ -253,6 +254,256 @@ public static StringValues Concat(StringValues values1, StringValues values2)
             return new StringValues(combined);
         }
 
+        public static bool Equals(StringValues left, StringValues right)
+        {
+            if (IsNullOrEmpty(left))
+            {
+                return IsNullOrEmpty(right);
+            }
+
+            if (left._values == null)
+            {
+                if (right._values == null)
+                {
+                    return left._value == right._value;
+                }
+
+                switch (right._values.Length)
+                {
+                    case 0:
+                        return left._values == null;
+                    case 1:
+                        return left._value == right._values[0];
+                    default:
+                        return false;
+                }
+            }
+
+            switch (left._values.Length)
+            {
+                case 0:
+                    return IsNullOrEmpty(right);
+
+                case 1:
+                    if (right._values == null)
+                    {
+                        return left._values[0] == right._value;
+                    }
+                    if (right._values.Length != 1)
+                    {
+                        return false;
+                    }
+                    return right._values[0] == left._values[0];
+
+                default:
+                    if (right._values == null)
+                    {
+                        return false;
+                    }
+                    if (left._values.Length != right._values.Length)
+                    {
+                        return false;
+                    }
+                    for (var i = 0; i < left._values.Length; i++)
+                    {
+                        if (left._values[i] != right._values[i])
+                        {
+                            return false;
+                        }
+                    }
+                    return true;
+            }
+        }
+
+        public bool Equals(StringValues other)
+        {
+            return Equals(this, other);
+        }
+
+        public static bool Equals(string left, StringValues right)
+        {
+            return Equals(right, left);
+        }
+
+        public static bool Equals(StringValues left, string right)
+        {
+            if (string.IsNullOrEmpty(right))
+            {
+                return IsNullOrEmpty(left);
+            }
+
+            if (IsNullOrEmpty(left))
+            {
+                return false;
+            }
+
+            if (left._values == null)
+            {
+                return left._value == right;
+            }
+
+            switch (left._values.Length)
+            {
+                case 1:
+                    return left._values[0] == right;
+
+                default:
+                    return false;
```

Comment:
Any reason with this + the `string[]` version not to construct a new `StringValues` and call that version of `.Equals`?